### ISSUE TITLE
ENH: Bump boost.math version

### ISSUE THREAD

**Issue by h-vetinari** [scipy/scipy#15136]
Breaking this out into a separate issue since the [comment](https://github.com/scipy/scipy/pull/13347#issuecomment-983295894) might otherwise easily get lost in a superseded PR:
> @mckib2: It occurs to me that we can try upgrading Boost now that we have a more recent minimum compiler that can handle more modern C++. That may help get upstream fixes for Boost.Math that we've been patching manually, might help with [gh-14901](https://github.com/scipy/scipy/issues/14901) and friends

In particular, now that the follow-up #15066 was merged, it's possible for SciPy to use C++14, which was probably the limiting factor for progressing further with boost?

**Comment by h-vetinari** [comment]
On the other hand, looking at the current status of the submodule in [`scipy/_lib`](https://github.com/scipy/scipy/tree/master/scipy/_lib), it looks from https://github.com/scipy/scipy/commit/5bbd0a26dbfd340816b48a236c8a9af6d4b34457 like we might need GCC 7.1, which is still gonna take a while...?

**Comment by rgommers** [comment]
> it looks from [5bbd0a2](https://github.com/scipy/scipy/commit/5bbd0a26dbfd340816b48a236c8a9af6d4b34457) like we might need GCC 7.1, which is still gonna take a while...?

This is unblocked now. Upgrading Boost sounds potentially useful, there's quite a few annoyances with build warnings and floating-point warnings that may at least in part go away.

**Comment by mckib2** [comment]
@rgommers Are all CI jobs now running GCC >=8?  Quick glance at [toolchain docs](https://docs.scipy.org/doc/scipy/dev/toolchain.html) says GCC 6 & 7, and azure-pipelines config is installing GCC 8 (but[ has a comment about GCC 6](https://github.com/scipy/scipy/blob/168eb5a4cea0d7ba5f8ec0e91045053d82947d12/azure-pipelines.yml#L116)).

Assuming GCC 8 (which seems to be lowest in gh-16589), I'll update the boost headers-only repo

**Comment by h-vetinari** [comment]
I haven't had time to lift all CI jobs after the minimum was agreed to be lifted to GCC 8. Some changes in this direction have happened already AFAIK, but the toolchain docs are likely not up to date on that. I'll try to have a look next week if no-one else gets there before.

**Comment by mckib2** [comment]
Just for fun and to see what happens, I created gh-17207 which builds with the latest version of Boost

The [release docs](https://www.boost.org/users/history/version_1_80_0.html) suggest this might work:
![image](https://user-images.githubusercontent.com/28734319/195210312-620e02ac-c8b5-4c8c-8aa9-a99e5eac9cab.png)

**Comment by rgommers** [comment]
The docs must be needing a tweak, the main `meson.build` will raise a hard error for GCC < 8.
